### ISSUE TITLE
Re-bgzip files according to .gzi

### ISSUE THREAD

**Issue by jacrayner** [samtools/htslib#377]
Hi, I would like to be able to bgzip according to a bgzip index file. 
The issue I have is that different implementations of bgzip segregate the uncompressed data into blocks in different ways. For instance if I convert a BAM to CRAM and back to BAM, will not necessarily  produce a identical BAM due to bgzip compression. Compression level/strategy is easy enough to match but  block segregation is less so.

The index file of bgzip files contains uncompressed offset and compressed offset for each block. The uncompressed offset essentially tells you how much uncompressed data is stored in each block and thus how to segregate the uncompressed data when re-bgzipping to match a previously bgzipped file.  

I have achieved this by adding another bgzip_write function to bgzf.c which just writes the data you give it into a block each time it is called. 

 ssize_t bgzf_block_write(BGZF *fp, const void *data, size_t length)
{
    if ( !fp->is_compressed )
        return hwrite(fp->fp, data, length);
    const uint8_t *input = (const uint8_t*)data;
    assert(fp->is_write);
    uint8_t* buffer = (uint8_t*)fp->uncompressed_block;
    memcpy(buffer, input, length);
    fp->block_offset += length;
    if (lazy_flush(fp) != 0) return -1;
    return length;
}

And then I added corresponding options and logic to bgzip.c to parse an index file and then feed uncompressed data to this function based on the index. 

Could I ask for comments whether something like this could be included in bgzip and any advice on best implementation if you seen any problems in mine? 

Thanks

**Comment by jkbonfield** [comment]
I can see your point, but I fear it's never going to work.  There could be many reasons why BAM to CRAM to BAM can yield different data, not only due to boundaries but potentially also different size for aux elements, perhaps different order in aux tags, headers having M5 entries, etc.  Plus there could be a myriad of different ways the starting BAM was produced - possibly from Picard or maybe from Samtools; did it always start a BGZF block on a new sequence boundary, etc.  Put simply, we cannot guarantee the uncompressed BAM is bit level identical so keeping the index identical is inherently going to be problematic.

What problem precisely are you trying to solve here?  Why does it matter that the index is identical?  Perhaps CRAM -> BAM producing both BAM and BAM.bai simultaneously is a better solution (I'm not sure if it's possible already).

**Comment by jkbonfield** [comment]
Infact I just did a test demonstrating this.

```
$ java -Xmx4g -jar SamFormatConverter.jar I=1m.bam O=1mp.bam COMPRESSION_LEVEL=0
$ samtools view -h 1mp.bam |samtools view -u > 1ms.bam
$ zcat 1mp.bam |md5sum
d48051e53083326771401d7245cf1e2e  -
$ zcat 1ms.bam |md5sum
3f48894b9f33a1a17d533f5018944ccc  -
```

So using htsjdk to generate a BAM file and then converting this from BAM -> SAM -> BAM in samtools again gives us a different BAM after stripping out any bgzf block boundary differences.

The difference turns out to be choice of *C* vs *c* type for the one byte X0 tag (amongst others).  While the uncompressed data is identical size, the choice of aux type means the bgzf output differs (even if I uncompress and recompress it using an identical bgzip binary).  If we cannot guarantee bit level identity of BAM->SAM->BAM we have zero chance of this happening in a CRAM world.

**Comment by jacrayner** [comment]
Yes I can see your point also. Essentially my issue is one of perception more than what is technically necessary. I am validating the use of cram in a clinical setting and as such require the ability to retrieve a bit identical file; at least during the validation stage. 

This seems to be achievable with the pipeline I am using as I have validated this process on around 700 bams thus far. However I can see that there are a number of caveats that means it will not be a generally achievable solution across all tools:

- Requires the use of scramble with the -P -p options.
- BAM requires m5 header entries already present.
- Order of the BAM header lines need to be resorted.
- Bgzip of aligner and scramble vary, thus the need for rebgzip and the bgzip index 
- Need to record the BGZIP compression level and strategy somewhere.
- Other issues that will inevitably arise.

So I can see your point that such a feature of bgzip may be quite niche.

**Comment by jkbonfield** [comment]
I just twigged who you are, sorry the username should have been obvious!  Of course I realise now, given the changes I made to scramble.

Given bgzip is a transparent layer, is it not sufficient to demonstrate that zcat on the before/after files are identical and therefore the bgzipped files contain identical content?  If however there is absolutely no way you can accept that identical content means identical files and MUST have bit level compressed BAM then personally I don't see it as a big deal to extend bgzip some more if the changes are small enough, although I'm just one voice and we'd need to discuss it I think.

For a more general approach you need some of the same features present in Precomp (https://github.com/schnaader/precomp-cpp) in that you may need to analyse what Deflate parameters were initially used so you can reuse them.  However for a single validated pipeline you control all points so it's an easier task.

**Comment by jacrayner** [comment]
ah yeah sorry James I realise we have had similar conversations already and you have been a considerable help. 

Unfortunately yes getting a bit match on the compressed data seem to be the point of least resistance.

Thanks for the advice that would be useful but as all I have to go on from the current bgzf index is the size of the compressed and uncompressed data, I imagine that the solution will have to be: feed in the same amount of uncompressed data and check that the correct size of compressed data is produced. I can also see that I should be calling `bgzf_index_load()` to parse the index file to do this in a constant way.
